maas/resource_maas_boot_source_test.go

```diff
@@ -0,0 +1,101 @@
+package maas_test
+
+import (
+	"fmt"
+	"strconv"
+	"terraform-provider-maas/maas/testutils"
+	"testing"
+
+	"github.com/canonical/gomaasclient/client"
+	"github.com/canonical/gomaasclient/entity"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+func TestAccResourceMAASBootSource_basic(t *testing.T) {
+
+	var bootsource entity.BootSource
+	url := "http://images.maas.io/ephemeral-v3/candidate/"
+
+	checks := []resource.TestCheckFunc{
+		testAccMAASBootSourceCheckExists("maas_boot_source.test", &bootsource),
+		resource.TestCheckResourceAttr("maas_boot_source.test", "url", url),
+	}
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testutils.PreCheck(t, nil) },
+		Providers:    testutils.TestAccProviders,
+		CheckDestroy: testAccCheckMAASBootSourceDestroy,
+		ErrorCheck:   func(err error) error { return err },
+		Steps: []resource.TestStep{
+			{
+				Config: testAccMAASBootSource(url),
+				Check:  resource.ComposeAggregateTestCheckFunc(checks...),
+			},
+		},
+	})
+}
+
+func testAccMAASBootSourceCheckExists(rn string, bootSource *entity.BootSource) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[rn]
+		if !ok {
+			return fmt.Errorf("resource not found: %s\n %#v", rn, s.RootModule().Resources)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("resource id not set")
+		}
+
+		conn := testutils.TestAccProvider.Meta().(*client.Client)
+		id, err := strconv.Atoi(rs.Primary.ID)
+		if err != nil {
+			return err
+		}
+		gotBootSource, err := conn.BootSource.Get(id)
+		if err != nil {
+			return fmt.Errorf("error getting boot source: %s", err)
+		}
+
+		*bootSource = *gotBootSource
+
+		return nil
+	}
+}
+
+func testAccMAASBootSource(url string) string {
+	return fmt.Sprintf(`
+resource "maas_boot_source" "test" {
+	url = "%s"
```

Comment:
you should also set a keyring file here. Or just use the default one. Otherwise, the tests will fail since when you first create, the provider state will get the keyring information from MAAS during read. So, after the appliance with absent keyring file name, the provider will think that user wants to unset the keyring file name. Which is not the case.